llvm/lib/Target/Mips/MipsISelLowering.cpp

```diff
@@ -876,38 +876,84 @@ static SDValue performANDCombine(SDNode *N, SelectionDAG &DAG,
 static SDValue performORCombine(SDNode *N, SelectionDAG &DAG,
                                 TargetLowering::DAGCombinerInfo &DCI,
                                 const MipsSubtarget &Subtarget) {
-  // Pattern match INS.
-  //  $dst = or (and $src1 , mask0), (and (shl $src, pos), mask1),
-  //  where mask1 = (2**size - 1) << pos, mask0 = ~mask1
-  //  => ins $dst, $src, size, pos, $src1
   if (DCI.isBeforeLegalizeOps() || !Subtarget.hasExtractInsert())
     return SDValue();
 
-  SDValue And0 = N->getOperand(0), And1 = N->getOperand(1);
+  SDValue FirstOperand = N->getOperand(0), SecondOperand = N->getOperand(1);
   unsigned SMPos0, SMSize0, SMPos1, SMSize1;
   ConstantSDNode *CN, *CN1;
+  uint64_t Pos = 0;
+
+  if ((FirstOperand.getOpcode() == ISD::AND &&
+       SecondOperand.getOpcode() == ISD::SHL) ||
+      (FirstOperand.getOpcode() == ISD::SHL &&
+       SecondOperand.getOpcode() == ISD::AND)) {
+    // Pattern match INS.
+    //   $dst = or (and $src1, (2**size0 - 1)), (shl $src2, size0)
+    //   ==> ins $src1, $src2, pos, size, pos = size0, size = 32 - pos;
+    //   Or:
+    //   $dst = or (shl $src2, size0), (and $src1, (2**size0 - 1))
+    //   ==> ins $src1, $src2, pos, size, pos = size0, size = 32 - pos;
+    SDValue Operand1 = FirstOperand.getOpcode() == ISD::AND
+                           ? FirstOperand.getOperand(1)
+                           : SecondOperand.getOperand(1);
+    if (!(CN = dyn_cast<ConstantSDNode>(Operand1)) ||
+        !isShiftedMask_64(CN->getZExtValue(), SMPos0, SMSize0))
```

Comment:
There has a check `bool hasExtractInsert() const { return !inMips16Mode() && hasMips32r2(); }` at the beginning of the function, then we don’t need to add the judgment you mentioned?

Reply:
No. `'mips64r2'.hasMips32r2()` also returns `True`.
This test is not about ABI.

Reply:
Sorry. I mean the VT of `AndOperand` or `ShlOperand`. Not O32/N32/N64.

Reply:
Yes, I got it. Because other combine related functions all use `isShiftedMask_64`, so I dod not consider `isShiftedMask_32` before. I would update it.

Reply:
Should we change other original `isShiftedMask_64` in this function `performORCombine`?

Reply:
When using `isShiftedMask_64`, its parameter has been zero expanded with `CN->getZExtValue()`, is that okay?

Reply:
Ohh. We have `assert(SMPos1 < ValTy.getSizeInBits());`, so anyway, there won't something wrong code be generated.